### ISSUE TITLE
Question about rule-based task generators (Make-like)

### ISSUE THREAD

**Issue by hazelnusse** [waf-project/waf#2114]
What would I need to change to make two (or more) rules similar to the following guarantee correct execution order? I often have tasks to do that are most simply done using Unix command line tools but I haven't figured out a simple way to ensure that these types of rules are executed in the correct order. For example:
```
top = '.'
out = 'build'

def configure(conf):
        pass

def build(bld):
    bld(rule = 'cp ${SRC} ${TGT}', source = 'wscript', target = 'foobar.txt')
    bld(rule = 'cp ${SRC} ${TGT}', source = 'foobar.txt', target = 'foobah.txt')
```

**Comment by ita1024** [comment]
The order is already guaranteed by the rather explicit file chain: `wscript` -> `foobar.txt` -> `foobah.txt`

**Comment by hazelnusse** [comment]
If the first rule is generating a C header file .h from some template, and the second rule is building a program that needs to include that generated header, when I specify the `--targets=` for the second target, the first target is not created before the second rule is run, and the compilation fails. For example, if I have this:

`def build(bld):
    bld(rule = 'cp ${SRC} ${TGT}', source = 'dummy.h.in', target = 'dummy.h')
    bld.program(target = 'footest', source = 'main.c', includes='.')
`
How do I indicate that `footest` must depend upon `dummy.h`? Listing 'dummy.h' in the footest sources doesn't seem to work either.

**Comment by ita1024** [comment]
You have changed the question.

1. for the explicit task order declaration, read the `ext_out` part of https://github.com/waf-project/waf/blob/master/demos/c/program/wscript_build#L5 (unnecessary in many cases as long as the output files and the task scanners are written correctly)
2. for bringing additional targets with --targets, see the "use" system (`XYZ` in the example) or build groups.

In both cases, read The Waf Book.

**Comment by hazelnusse** [comment]
Yes, sorry for having changing the question.
I have read the entire waf book, multiple times. Despite having done this, multiple times, it was/is not obvious to me the right approach, hence the question. I hope me asking questions here is welcome -- if not, please let me know where such questions should be asked.

**Comment by hazelnusse** [comment]
So, if I understand correctly, the following approach is the correct way to achieve this:

```
def build(bld):
    bld(features='use',
        rule = 'sleep 1 && cp ${SRC} ${TGT}',
        source = 'dummy.h.in',
        target = 'dummy.h',
        ext_out=['.h'])
    bld.program(target = 'footest', source = 'main.c', includes='.')
```

It seems to also work correctly even when I don't use the `feature='use'` and the `ext_out=['.h']` arguments -- presumably this is because the scanners that detect that `main.c` includes the `dummy.h`?

**Comment by ita1024** [comment]
You forgot the **XYZ** part in the example (`--target=...`).

The `ext_out` part is usually best to keep for documentation purposes and for performance reasons. Read the outputs of `waf clean build --zones=deps`.

**Comment by hazelnusse** [comment]
I worked on this a bit more and it seems that the behavior is different between generated .c vs .h files. I've created a set of minimal examples here:

https://github.com/hazelnusse/waf_sandbox

I'm able to make a generated .h file work with and without the --targets, but presumably this is due to a scanner that is detecting this dependency in the .c file that is including the generated file.

I can't get the same type of behavior from a generated .c file which is my actual use case. Any suggestions on what is needed to make the test0 and test1 examples work?
